File: py/0098_Validate_Binary_Search_Tree.py
import sys
import heapq
from typing import *
from collections import OrderedDict
from collections import deque


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def helper(self, root, l, r):
        if not root:
            return True

        if l >= root.val or r <= root.val:
            return False

        return self.helper(root.left, l, root.val) and self.helper(root.right, root.val, r)

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.helper(root, -sys.maxsize, sys.maxsize)


if __name__ == "__main__":
    solution = Solution()

    root = TreeNode(2)
    root.left, root.right = TreeNode(1), TreeNode(3)
    print(solution.isValidBST(root))

    root = TreeNode(5)
    root.left, root.right = TreeNode(1), TreeNode(4)
    root.right.left, root.right.right = TreeNode(3), TreeNode(6)
    print(solution.isValidBST(root))

    sys.exit(0)

File: cpp/0098_Validate_Binary_Search_Tree.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <list>
#include <queue>
#include <stack>
#include <map>
// #include <limits>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool check(TreeNode* root, TreeNode* l, TreeNode* r) {
        if (!root) return true;
        if (l && root->val <= l->val) return false;
        if (r && root->val >= r->val) return false;

        return check(root->left, l, root) && check(root->right, root, r);
    }

    bool isValidBST(TreeNode* root) {
        return check(root, nullptr, nullptr);
    }
};

/*
class Solution {
public:
    int last_val;
    bool set;

    bool dfs(TreeNode* root) {
        if (root->left) {
            if (!dfs(root->left)) return false;
        }

        if (last_val >= root->val) {
            if (set)
                return false;
        }
        last_val = root->val;
        set = true;

        if (root->right) {
            if (!dfs(root->right)) return false;
        }

        return true;
    }

    bool isValidBST(TreeNode* root) {
        last_val = numeric_limits<int>::min();
        set = false;

        return dfs(root);
    }
};
*/

int main(int argc, char *argv[]) {

    return 0;
}
